### ISSUE TITLE
REPL crash in v1.5.3 with `f(0u"s") = 0`

### ISSUE THREAD

**Issue by icantbelieveitsnotphysics** [JuliaLang/julia#39656]
The following code entered into the REPL

```julia
using Unitful
f(x) = 0
f(0u"s") = 0
```

causes the process to crash with a `EXCEPTION_ACCESS_VIOLATION` under julia v1.5.3 on windows 10 (installed via scoop) and Unitful v1.5.0 (1986cc42).

Versioninfo:

```txt
Julia Version 1.5.3
Commit 788b2c77c1 (2020-11-09 13:37 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i5-7500 CPU @ 3.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
```

Crash spew: [juliacrash.txt](https://github.com/JuliaLang/julia/files/5978002/juliacrash.txt)

I also checked on v1.4.1 on a raspberry pi (exactly the same Unitful version), and the above code did _not_ crash but instead raised the following error:

```julia-repl
ERROR: syntax: "(0 * Unitful.FreeUnits{(Unitful.Unit{:Second, Unitful.Dimensions{(Unitful.Dimension{:Time}(power=Base.Rational{Int32}(num=1, den=1)),)}()}(tens=0, power=Base.Rational{Int32}(num=1, den=1)),), Unitful.Dimensions{(Unitful.Dimension{:Time}(power=Base.Rational{Int32}(num=1, den=1)),)}(), nothing}())" is not a valid function argument name
```

This slightly unhelpful message is the expected response under windows (and seems to have been the standard error in these circumstances for some years, see #15421).

Versioninfo:
```text
Julia Version 1.4.1
Commit 381693d3df* (2020-04-14 17:20 UTC)
Platform Info:
  OS: Linux (arm-linux-gnueabihf)
  CPU: ARMv7 Processor rev 3 (v7l)
  WORD_SIZE: 32
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, cortex-a72)
```

Given the nature of the crash I'm assuming this is a Julia issue rather than a Unitful issue, but I don't know enough about either to be certain. I've been unable to force the crash to occur with argument types other than a Unitful quantity, including regex string macros, but I haven't tried many different things.

**Comment by sostock** [comment]
The crash in 1.5.3 occurs also under linux-x86_64. It does not happen on 1.4.2, 1.6-rc1 and master (at least for me).

The error message is indeed expected. Is it just the verbose printing of `0u"s"` that you consider unhelpful in the error message?

**Comment by icantbelieveitsnotphysics** [comment]
@sostock its more that "_not a valid function argument name_" isn't very helpful when I would have expected something like "_you are trying to assign a value to something that can't take one_".

Verbose printing of Unitful types does give me unpleasant flashbacks to C++ templates, but at least they are theoretically meaningful.

**Comment by sostock** [comment]
> its more that "_not a valid function argument name_" isn't very helpful when I would have expected something like "_you are trying to assign a value to something that can't take one_".

The way I see it, `f(something) = 0` is not assignment syntax, but method definition syntax. When the left-hand side is not of the form `f(something)`, an error similar to the one you propose is actually thrown:

```julia
julia> f(x) = 0 # defining a method
f (generic function with 1 method)

julia> x = 0 # assigning a value
0

julia> f(5) = 0 # trying to define a method, but with an invalid argument
ERROR: syntax: "5" is not a valid function argument name around REPL[4]:1

julia> 5 = 0 # trying to assign a value: basically the error message you propose
ERROR: syntax: invalid assignment location "5" around REPL[5]:1
```

**Comment by sostock** [comment]
The crash does also not happen on 1.5.0, so the bug was introduced between 1.5.0 and 1.5.3.

**Comment by sostock** [comment]
Bisect says this was introduced by 58cd4b887cc63584d399c553694cfbd355b0c852.

**Comment by KristofferC** [comment]
Fixed in the upcoming 1.6:

```jl
julia> using Unitful

julia> f(x) = 0
f (generic function with 1 method)

julia> f(0u"s") = 0
ERROR: syntax: "(0 * Unitful.FreeUnits{(Unitful.Unit{:Second, Unitful.Dimensions{(Unitful.Dimension{:Time}(power=Base.Rational{Int64}(num=1, den=1)),)}()}(tens=0, power=Base.Rational{Int64}(num=1, den=1)),), Unitful.Dimensions{(Unitful.Dimension{:Time}(power=Base.Rational{Int64}(num=1, den=1)),)}(), nothing}())" is not a valid function argument name around REPL[7]:1
Stacktrace:
 [1] top-level scope
   @ REPL[7]:1
```

**Comment by icantbelieveitsnotphysics** [comment]
@sostock that's a good explanation, thanks.
